crates/runtime/src/instance/allocator/pooling/index_allocator.rs

```diff
@@ -30,56 +61,112 @@ pub(crate) enum PoolingAllocationState {
     /// the given module ID, if any. If not, we pick a random slot
     /// ID. This random choice is unbiased across all free slots.
     ReuseAffinity {
-        // Free-list of all slots. We use this to pick a victim when
-        // we don't have an appropriate slot with the preferred
-        // affinity.
-        free_list: Vec<usize>,
-        // Invariant: any module ID in this hashmap must have a
-        // non-empty list of free slots (otherwise we remove it).
-        per_module: HashMap<CompiledModuleId, Vec<usize>>,
-        // The state of any given slot. Records indices in the above
-        // list (empty) or two lists (with affinity), and these
-        // indices are kept up-to-date to allow fast removal.
+        /// Free-list of all slots. We use this to pick a victim when
+        /// we don't have an appropriate slot with the preferred
+        /// affinity.
+        free_list: Vec<SlotId>,
+        /// Invariant: any module ID in this hashmap must have a
+        /// non-empty list of free slots (otherwise we remove it). We
+        /// remove a module's freelist when we have no more slots with
+        /// affinity for that module.
+        per_module: HashMap<CompiledModuleId, Vec<SlotId>>,
+        /// The state of any given slot. Records indices in the above
+        /// list (empty) or two lists (with affinity), and these
+        /// indices are kept up-to-date to allow fast removal.
         slot_state: Vec<SlotState>,
     },
 }
 
 #[derive(Clone, Debug)]
 pub(crate) enum SlotState {
-    Taken,
-    Empty {
-        /// Index in the global free list. Invariant:
-        /// free_list[slot_state[i].free_list_index] == i.
-        free_list_index: usize,
+    /// Currently allocated.
+    ///
+    /// Invariant: no slot in this state has its index in either
+    /// `free_list` or any list in `per_module`.
+    Taken(Option<CompiledModuleId>),
+    /// Currently free. A free slot is able to be allocated for any
+    /// request, but may have affinity to a certain module that we
+    /// prefer to use it for.
+    ///
+    /// Invariant: every slot in this state has its index in at least
+    /// `free_list`, and possibly a `per_module` free-list; see
+    /// FreeSlotState.
+    Free(FreeSlotState),
+}
+
+impl SlotState {
+    fn as_free(&self) -> &FreeSlotState {
```

Comment:
I'd personally add `unwrap_` as a prefix to these match-on-a-variant-and-return-the-inner-bits-or-else-panic methods. Typically a plain `as_blah` would return an `Option<&T>` if it is possibly not a `T`, rather than panicking.

Reply:
Done!